### ISSUE TITLE
High availability support for  Federation

### ISSUE THREAD

**Issue by shaneu** [apollographql/federation#349]
First let me say we love federation on our team. We split our monolith into approx. 11 microservices and have been very pleased with the ease of the refactor and the results.

There were some issues around availability that we had to roll ourselves, and it led me to think maybe these would be better included in the federation/gateway package. 

First of all, the gateway attempts to connect to the federated services, and if it is unable it still get's initialized, only without the services it failed to connect too. This is troublesome for an application that is deployed automatically as it requires human intervention to monitor if the gateway initialization was successful. 

Our workaround was to create a `readinessProbe` that performs the same query the gateway makes to the federated services, with the goal being to make sure the services are available and ready to receive traffic. This probe has a configurable number of retries and timeout before throwing. If the probe is unsuccessful we let the application to crash, as it doesn't make sense for us to allow it to initialize without all services being available. 

Perhaps this feature could be built into the gateway, along with a configurable number of retries? If it is unsuccessful it crashes and does not init. Perhaps that too could be a configuration `mustConnect: true/false` (I'm sure there's a better name for the property).

Other than that we let kubernetes manage the availability of the other services via liveness probes, but for teams not using a deployment orchestration manager like a kubernetes perhaps there is a way by which the gateway could check the availability of its downstream services?

Also, a periodic reloading of the schema would be helpful. Let's take the case where we update a schema in one of our services and redeploy it. The gateway wouldn't know the service's schema had changed so we must restart the gateway for it to get the changes. This hurts availably by causing downtime while the gateway is redeployed. If the gateway would refresh its schema on a configurable polling interval that issue could be avoided. We've developed a hack using helm charts to tell the gateway to redeploy when one of the downstream services have been updated, but that still causes downtime. 

If any of these ideas seems like it would be a good addition to the project I would be more than happy to make a PR. I am truly appreciate of the work the apollo team is doing and I'd love to contribute.

**Comment by Nicoowr** [comment]
@rtymchyk Does the gateway immediately updates its configuration whenever the registry is updated in managed mode? I thought the gateway would poll every interval of time the registry, which could introduce short disruptions on schema changes.

**Comment by rtymchyk** [comment]
@Nicoowr It's not immediate, but it's close. It shouldn't matter too much though if your API evolution is always backwards compatible.

**Comment by Nicoowr** [comment]
@rtymchyk very clear thanks. As a side note, breaking changes happen from time to time in our case, hope this won't be a problem though.

**Comment by rhzs** [comment]
> Managed federation still doesn't re-initialiaze if it fails to pick up the schema. The polling timer never gets started so you just have a dud running service.

After a year, I don't see improvement on this one. I am curious why this is not highest priority / critical. Does one use federated GraphQL in production with high scale and not having this issue and blocker?

Can anyone point me best practices repo example to deploy my GraphQL services with CI/CD pipeline that accommodate re-initialisation if some services are offline?

**Comment by rtymchyk** [comment]
> > Managed federation still doesn't re-initialiaze if it fails to pick up the schema. The polling timer never gets started so you just have a dud running service.
> 
> After a year, I don't see improvement on this one. I am curious why this is not highest priority / critical. Does one use federated GraphQL in production with high scale and not having this issue and blocker?
> 
> Can anyone point me best practices repo example to deploy my GraphQL services with CI/CD pipeline that accommodate re-initialisation if some services are offline?

Are you talking about initial boot or mid lifecycle?

**Comment by glasser** [comment]
>>Managed federation still doesn't re-initialiaze if it fails to pick up the schema. The polling timer never gets started so you just have a dud running service.

>After a year, I don't see improvement on this one. I am curious why this is not highest priority / critical. Does one use federated GraphQL in production with high scale and not having this issue and blocker?


@rhzs The relevant change here is in Apollo Server. Starting with AS v2.22, the `listen()` method of `apollo-server` or the new `start()` method if you're using an integration like `apollo-server-express` will throw if the schema is not loaded properly on startup. If startup succeeds, the polling timer will get started and should run reliably (including if future attempts to update the schema have errors). There have also been some reliability improvements in Gateway itself around the polling logic, most notalby in Gateway 0.23.

**Comment by rtymchyk** [comment]
@glasser's change to `listen` has been working great in our Kubernetes environment :) No more dead pods on boot.

**Comment by kevinschaffter** [comment]
Apart from going down the path of managed federation, is there any way to prevent the entire GQL gateway from crashing if one of the federated services is not available?

**Comment by rhzs** [comment]
> Are you talking about initial boot or mid lifecycle?

Initial boot. mid lifecycle works (when service went down and back up again). 



> > > Managed federation still doesn't re-initialiaze if it fails to pick up the schema. The polling timer never gets started so you just have a dud running service.
> 
> > After a year, I don't see improvement on this one. I am curious why this is not highest priority / critical. Does one use federated GraphQL in production with high scale and not having this issue and blocker?
> 
> @rhzs The relevant change here is in Apollo Server. Starting with AS v2.22, the `listen()` method of `apollo-server` or the new `start()` method if you're using an integration like `apollo-server-express` will throw if the schema is not loaded properly on startup. If startup succeeds, the polling timer will get started and should run reliably (including if future attempts to update the schema have errors). There have also been some reliability improvements in Gateway itself around the polling logic, most notalby in Gateway 0.23.

`if start up succeeds`, does that mean we need to resolve all services first? IMAGINE, if I have 100s services in production. Does that make sense to wait 100s services UP and RUNNING in order the Gateway "initial startup" to succeeds? 

I have noticed the improvement on the pooling for handling service unavailability (Kudos for the team. ), but that only works after you success starting (startup) all the services.

**Comment by glasser** [comment]
Yes, your server can't start unless it knows its schema, which requires knowing all the subgraph schemas or being started directly with a supergraph schema.  Polling a bunch of individual subgraphs is certainly a big process which is part of why we've developed managed federation and other mechanisms of getting schemas into Gateway.

**Comment by rhzs** [comment]
@glasser that should be at least in this projectREADME / Apollo Docs about Apollo Federation scalability. Not much articles discussing about Apollo Federation scalability until I experienced it on my own.

**Comment by rhzs** [comment]
@glasser I am curious to know, can't we just check the GraphQL schema hash (using SHA) and cache the schema in local file, rather than fetched all remote schemas again, and only fetched the necessary remote schema (or maybe just part of it)?

**Comment by glasser** [comment]
Sure, that is something you could build with the primitives.

**Comment by j** [comment]
tldr; Please allow cached federated service configuration for HA.

@rhzs That makes way too much sense to do though. =P

Just like most GraphQL implementations allow you to emit a schema file (useful for development & having other libraries utilize schema.. i.e. client side generators, etc).

If Apollo Federation did something similar, that'd make a ton more sense.  Yes, maybe in development mode the gateway can fail if a service isn't live, but in production mode, if you can use a cached file describing the services to fall back on in case of a service being down, then to me that would be a whole lot better than the gateway failing.

Likely scenario: I have a gateway running on GCR (docker) and all my services running on GCR.  If a service goes down and GCR scales my gateway instances up, I'll end up with a bunch of failed docker containers when I'd rather have the service queue up requests and serve gateway timeout errors after a certain time.  Ideally still servicing all other service requests that are up.  I get that yeah we can have this entirely huge E2E test suite that runs on CI before going live, but sometimes we have to do a rushed push (configuration changes, etc).

Also, let's say that there's less mission critical services (admin analytics) and mission critical services (payment processing) and we break the admin analytics somehow or have downtime on that service.... our entire application is down?  We're not making money?

C'mon.  This is one of the bigger reasons people move to micro service architecture. I keep seeing "devil's advocate" responses pushing people to the managed services instead of actually realizing this can be an issue.

I worked all day on moving to Federation but think I'm going to go back to a hand-coded GraphQL "gateway" with gRPC micro-services because I ran into this issue and can see where it'll become a PITA in production and cause revenue loss.

/cc @rtymchyk @glasser

**Comment by j** [comment]
@rhzs I found a public attempt at re-creating a schema registry.

https://github.com/pipedrive/graphql-schema-registry

It's pretty much the route I was attempting (the example gateway).

Though they're extending ApolloGateway and recent changes to making everything private will likely break a library like this as the Apollo team doesn't want people to really extend ApolloGateway, etc (which I get, they just need to support things like this better and cleanly).

After playing with caching attempts, I'd like to figure out how to do the following (sudo-code):

```typescript
// ...

class ApolloGatewayFactory {
  protected gateway: ApolloGateway;
  protected pubsub: PubSub;

  async create(): Promise<ApolloGateway> {
    const { supergraphSdl } = await this.load();

    this.gateway = new ApolloGateway({ supergraphSdl });

    this.pubsub.subscribe('schema_changed', async () => this.update());

    return this.gateway;
  }

  async load(): Promise<SupergraphSdlGatewayConfig> {
    // .. get service from wherever you'd like, however you'd like
  }

  async update(): Promise<SupergraphSdlGatewayConfig> {
    const { supergraphSdl } = await this.load();

    this.gateway.update({ supergraphSdl });
  }
}

async start(): Promise<void> {
  const factory = new ApolloGatewayFactory();

  cost gateway = await factory.create();

  const server = new ApolloServer({
    gateway
  });

  composer.register(gateway);

  // ...
}
``

This is one example.  Here, I could build my own polling.  I could create an endpoint to force reload the gateway.  I could use etcd, consul, etc.  I could use managed federation but create a fallback for higher availability just in case managed Apollo sees downtime.

**Comment by kindermax** [comment]
There is another OSS registry https://github.com/StarpTech/graphql-registry. Just FYI

**Comment by j** [comment]
@kindermax That's cool too.  Good to see people see the issue.

There just needs to be better baked in support into `@apollo/*` libraries for 3rd parties to really utilize.  At any moment `graphql-registry` you mentioned can just break.  I did see that https://github.com/mercurius-js/mercurius implemented their own federation support using the spec.  Part of me feels sad that this is an option.

**Comment by j** [comment]
@kindermax Ironically, their examples already show similar code as what I mentioned:

https://mercurius.dev/#/docs/federation

```typescript
setTimeout(async () => {
  const schema = await server.graphql.gateway.refresh()

  if (schema !== null) {
    server.graphql.replaceSchema(schema)
  }
}, 10000)
```

**Comment by Borduhh** [comment]
Looking at how `IntrospectAndCompose` works, would it be easiest to just implement individual retry logic when it attempts to retrieve a data source?

Here is the code behind that:
```ts
return dataSource
      .process({
        kind: GraphQLDataSourceRequestKind.LOADING_SCHEMA,
        request,
        context: {},
      })
      .then(({ data, errors }): ServiceDefinition => {
        if (data && !errors) {
          const typeDefs = data._service.sdl as string;
          const previousDefinition = serviceSdlCache.get(name);
          // this lets us know if any downstream service has changed
          // and we need to recalculate the schema
          if (previousDefinition !== typeDefs) {
            isNewSchema = true;
          }
          serviceSdlCache.set(name, typeDefs);
          return {
            name,
            url,
            typeDefs: parse(typeDefs),
          };
        }

        throw new Error(errors?.map((e) => e.message).join('\n'));
      })
      .catch((err) => {
        const errorMessage =
          `Couldn't load service definitions for "${name}" at ${url}` +
          (err && err.message ? ': ' + err.message || err : '');

        throw new Error(errorMessage);
      });
```

**Comment by trevor-scheer** [comment]
@shaneu I believe the gateway now behaves in the way you'd like (fails to start on init if subgraphs are unavailable). Additionally, the somewhat recent introduction of the `supergraphSdl` field (aka `SupergraphManager`s) permits a lot of flexibility in userland code as well (see https://github.com/apollographql/federation/pull/1246 for more details).

I see a handful of other concerns in this issue which I believe are also addressed by `SupergraphManager`s, but if not I'd welcome anyone to open a separate issue with a specific ask.

@Borduhh that sounds reasonable, would you mind opening a separate issue for discussion?

Thanks all. I'll gladly reopen this issue if the original issue is in fact not resolved.
